src/main/kotlin/org/misarch/discount/service/DiscountService.kt

```diff
@@ -241,26 +241,99 @@ class DiscountService(
     }
 
     /**
-     * Finds all discounts that are applicable to a product variant, a user and a list of coupons.
+     * Finds the remaining usages for a user for a set of discounts
+     * Only returns the remaining usages for discounts that have a maximum number of usages per user
+     *
+     * @param userId the id of the user
+     * @param discounts the discounts to find the remaining usages for
+     * @return the remaining usages for each discount that has a maximum number of usages per user
+     */
+    private suspend fun findRemainingUsagesForDiscounts(
+        userId: UUID, discounts: Set<DiscountEntity>
+    ): Map<UUID, Long> {
+        val discountIds = discounts.filter { it.maxUsagesPerUser != null }.map { it.id!! }.toSet()
+        val currentUsagesByDiscount = discountUsageRepository.findByUserIdAndDiscountIdIn(userId, discountIds)
+            .associateBy({ it.discountId }) { it.usages }
+        return discounts.associate {
+            it.id!! to it.maxUsagesPerUser!! - (currentUsagesByDiscount[it.id] ?: 0)
+        }
+    }
+
+    /**
+     * Finds all discounts that are applicable to a user, and a list of product variant and a list of coupons.
      * Also checks if the user can use the discounts with the amount of items.
      *
      * @param input the input for the query
-     * @return The applicable discounts
+     * @return The applicable discounts for each product variant in order
```

Comment:
Ah, now I see why the output is so weird.
Still, wouldn't it make sense to remove the ordering and return a map instead?
Otherwise, the output is really hard to evaluate.
Or do we need the ordering?

Reply:
the problem is, you can provide multiple inputs for the same productVariant (at least semantically for this service)
I could forbid that, but that's also an arbitrary limitation
so afaik the order is the only way to be sure without an artificial key
however I will change it to a "map" (basically a list of tupels), because then if the caller does not provide duplicate product variants, the id can be used instead (and even more important, lists of lists draw unwanted code reviewer attention)